KVM: Rudimentary support for VCPI

bartoldeman opened the pull request:
Implement basic VCPI support with limitations.
The limitations are:
$_dpmi = (0)
$_ems = (0)
* don't use FDPP
* use external XMS provider (HIMEM/FDXMS)

VCPI clients allocate memory from XMS only, so EMS needs to report 0 memory, and XMS needs to map identical "physical" memory starting at 0x110000, which would normally be taken by DPMI memory.

The page table maps LOW+HMA as usual, then a page at 0x110000 to monitor code implementing a monitor->VCPI client jump and a VCPI client -> VM86 jump. A page at 0x111000 contains the saved monitor GDTR/IDTR/CR3 values and a temporary stack, with hiword(esp)=0, for the client to use. The monitor code is much like Jemm's.

When KVM interrupts the VCPI client through a signal, DOSEMU does its regular things, but CANNOT modify any registers, as the client registers stay in the VM. pic_run() will see VIF is not set, so won't modify, and the kvm.c code needs to use KVM_INTERRUPT if pic_pending().
The only other place that has been adapted for a callback into vm86 is leavedos().

VGAEMU memory is unprotected for now, so updates aren't reliable, and VGA planar modes don't work.

DOOM is playable but a bit slow due to frequent VGA I/O port accesses which would normally go via instremu. Duke Nukem 3D runs but with choppy sound and incomplete screen updates.

bartoldeman committed:
KVM: Rudimentary support for VCPI

Implement basic VCPI support with limitations.
The limitations are:
$_dpmi = (0)
$_ems = (0)
* don't use FDPP
* use external XMS provider (HIMEM/FDXMS)

VCPI clients allocate memory from XMS only, so EMS needs to report 0 memory,
and XMS needs to map identical "physical" memory starting at 0x110000, which
would normally be taken by DPMI memory.

The page table maps LOW+HMA as usual, then a page at 0x110000 to monitor code
implementing a monitor->VCPI client jump and a VCPI client -> VM86 jump.
A page at 0x111000 contains the saved monitor GDTR/IDTR/CR3 values and
a temporary stack, with hiword(esp)=0, for the client to use. The monitor code
is much like Jemm's.

When KVM is interrupts the VCPI client through a signal, DOSEMU does its regular
things, but CANNOT modify any registers, as the client registers stay in the VM.
pic_run() will see VIF is not set, so won't modify, and the kvm.c code needs
to use KVM_INTERRUPT if pic_pending().
The only other place that has been adapted for a callback into vm86 is leavedos().

VGAEMU memory is unprotected for now, so updates aren't reliable, and VGA planar
modes don't work.

DOOM is playable but a bit slow due to frequent VGA I/O port accesses which would
normally go via instremu. Duke Nukem 3D runs but with choppy sound and incomplete
screen updates.

bartoldeman committed:
VCPI: Quick fix to get Duke Nukem 3D working in 320x200

By dirtying all pages. Can be improved using KVM_GET_DIRTY_LOG.

stsp commented:
> VCPI clients allocate memory from XMS only, so EMS needs to report 0 memory, and XMS needs to
> map identical "physical" memory starting at 0x110000, which would normally be taken by DPMI memory.

I reverted that optimization in #1881.
Please let me know if this is what you
are looking for - that should allow vpci
to work with internal XMS.

Other than that - quite interesting, but
could you please describe exactly what
does this give us? win95?
Also you likely want to ruin the arm port,
don't you? :) Or maybe you want to move
to tcg, allowing the arm port to provide
vcpi too?

stsp commented:
Have you considered the use of
the KVM_CAP_SYNC_REGS, as
in #1408 ?
Unless I am missing the actual
problem you are facing, with
SYNC_REGS the aforementioned
limitations will go away because
you will always have the right
registers at hands. Monitor then
doesn't have to save/restore them.

stsp commented:
> VGAEMU memory is unprotected for now, so updates aren't reliable, and VGA planar modes don't work.

We still have the vgacpy branch in #1099 
Hmm, looks like we have many solutions
around, waiting for a problem. :) And
here it is, a problem.

stsp commented:
> The page table maps LOW+HMA as usual, then a page at 0x110000 to monitor code

Can you put it somewhere upwards,
into the "main_pool"?

stsp commented:
> By dirtying all pages. Can be improved using KVM_GET_DIRTY_LOG.

As in #198 ?

bartoldeman committed:
KVM VCPI: implement dirty page logging for VGA buffer

Much like in #198, use the KVM API to track which VGA pages have been written to.

bartoldeman commented:
At the moment this is mostly a playground to see what is possible, with a VCPI client having full control over the VM, and to test some ideas, perhaps find some bugs, it also allows quick checking how DPMI provider X behaves vs. dosemu2's DPMI implementation. Then some of the ideas could go into devel before this would be merged.

This is for the moment a KVM only thing, nothing against Arm, just a bonus when KVM is there for now.

Sure Windows 95 and various games that need VCPI or flat real mode would be a bonus for sure.


> >  The page table maps LOW+HMA as usual, then a page at 0x110000 to monitor code

> Can you put it somewhere upwards,
> into the "main_pool"?

How do you mean, moving the code and data page from the monitor to two pages allocated from the main pool?
I mean the whole monitor can be allocated from the main pool if we like now. Then we'd only have a single mmap for DOS addressable space.

VCPI spec says they must be mapped into the first 4MB (linear), could be in low space (e.g. I could put the VCPI<->VM86 switch code in bios.S), or mapped from somewhere high (which is what I'm doing now, monitor is high)

Note that before VCPI, KVM always identity maps, from guest virtual space to guest physical space. This keeps it consistent with native DPMI. But with VCPI the client has control over page tables so we only have control over physical space except (collaboratively!) for the first 1MB + 64k + up to 3 MB of virtual address space. That's why DPMI had to be disabled in the config, so that extmem is at physical (not linear) 0x110000.

> As in https://github.com/dosemu2/dosemu2/pull/198 ?
yes

stsp commented:
> That's why DPMI had to be disabled in the config, so that extmem is at physical (not linear) 0x110000.

This part I don't understand.
dpmi_base is far above ext_mem.
They are quite unrelated.
With the reverts I did in a nearby
branch, ext_mem is going to be
mapped to the physical 0x110000
under KVM by the means of EPT.
So what is the problem?

stsp commented:
ext_mem is aliased to some high space in
main_pool, and to 0x110000. Under KVM,
both these windows actually represent the
physical addresses. And the linear addresses
too, if we set up the identity page tables.
But I don't see any connection with dpmi,
could you please clarify?

bartoldeman commented:
oh I may be slightly out of date, as I see now that dpmi_base's default was moved from 8MB to 32MB. Which leaves max 31MB for extmem now.

Still for HX, 0x400000 (at 4MB linear) needs to be available for DPMI right, see https://github.com/dosemu2/dosemu2/issues/612?
So there's a conflict with extmem (if it uses more than 3MB, and mapped at 0x110000 linear) or is this handled somehow?

stsp commented:
Yeah, that's right, and that part
I forgot to revert. :) Added the missing
reverts now.
Also I am not sure ext_mem is
actually mapped where it should,
even with reverts, because
MAPPING_INIT_LOWRAM alias
still seems to be done w/o ext_mem
size in low_mem_init(). But that's
trivial to fix.
So I suppose perhaps you don't
need to disable dpmi fully, just break
HX when vcpi switched on?

stsp commented:
I added the patch that is supposed
to map it to kvm's phys addr now for
real. :)

stsp commented:
> So there's a conflict with extmem (if it uses more than 3MB, and mapped at 0x110000 linear) or is this handled 
> somehow?

In devel this is handled by mapping
ext_mem to higher addresses. Which
is the problem for vcpi, as under kvm
that "higher address" is a phys address.
In xms_rv branch this is handled by
lowering the ext_mem size...
But in a guest we don't need to map
ext_mem to any linear address. So perhaps
the better fix would be to leave things
as they are in devel and invent non-identity
maps for kvm? But that looks too tricky.

stsp commented:
Well I guess the easiest solution for
now would be for you to just integrate
the xms_rv branch into yours, w/o pushing
it to devel. This will allow you to play with
3Mb of ext_mem, or increase it by the
cost of breaking HX, and later come up
with some other solution.

stsp commented:
Hmm, perhaps I've found the simple
solution:
https://github.com/dosemu2/dosemu2/pull/1881/commits/57c667bf080002ebf1b137a74797c8cf1ac88a61
Why not to just use dpmi_rsv_low
as ext_mem?

stsp commented:
Except that then we shouldn't
write-protect reserved area, so
https://github.com/dosemu2/dosemu2/pull/1881/commits/1e62ebffec042890e5587d8d75884ae49736e279

stsp commented:
xms_rv built and I think its quite
simple to apply to devel. Would
you like to check that it actually
helps vcpi?

bartoldeman committed:
KVM: VCPI: handle planar VGA modes, slowly

Don't map the physical memory for the VGA buffer then so that all accesses
cause an MMIO KVM exit. instremu isn't read for paged memory, and probably
this will be fast enough with coalesced MMIO.

bartoldeman committed:
KVM VCPI: use coalesced MMIO

Gives much better speed in VGAEMU planar modes.

bartoldeman committed:
KVM VCPI: use coalesced PIO for VGA I/O in planar modes.

stsp commented:
I merged identity mapping for
ext_mem, so hope disabling DPMI
will not be needed for VCPI.

bartoldeman commented:
This definitely makes this easier though now I need to introduce config.vcpi obviously :)
Will do so in a few days when I have time.

stsp commented:
Oh would you like to complete the simx86
and fpu activities maybe?
Why vcpi can't wait? :)

bartoldeman committed:
KVM: VCPI: Eliminate config.dpmi checks except for vgaemu

This allows command line executables to run properly with VCPI and non-zero
config.dpmi.

For DJGPP that means running cwsdpmi explicitly, and for DOS4GW,
SET DOS16M=11

Still needs $_ems=(4), a low enough value so that EMS is enabled but
doesn't let VCPI allocate any memory from it.

bartoldeman committed:
KVM VCPI: make vgaemu independent of config.dpmi.

So that graphics work with VCPI where $_dpmi is not (0).

jharrison03 commented:
> DOOM is playable but a bit slow due to frequent VGA I/O port accesses which would normally go via instremu. Duke Nukem 3D runs but with choppy sound and incomplete screen updates.

Doom runs very slow.  Duke 3D actually runs very well for me.

stsp commented:
doom is slow and duke is ok even without
this patch. vcpi hardly makes a difference
here.

stsp commented:
So there seems to be some "market" for
this actually, eg #875 
It is about run286, which would be quite
difficult to run on our DPMI impl, even if
the reverse-engineered stuff from dosbox
exist. It relies on a particular descriptors
layout (expects LDT/IDT in certain slots
IIRC), and has crippled int21 translation.

Another thing is myjemm - ems extensions.
Should be much simpler to implement if
anyone ever cared.

With VCPI, both would hopefully just work.

bartoldeman committed:
KVM: VCPI: make it compile and run with present devel code

bartoldeman commented:
Still very much a draft, but a little less restrictive and more functional:
* `$_dpmi` no longer needs to be `(0)` but in that case DOS progs need to be told explicitly to use VCPI if you want them to (e.g. running them explicitly with `cwsdpmi.exe` (DJGPP) or `set DOS16M=11` (DOS4GW))
* Still needs `$_ems=(4)` an artificial low value, so EMS int67 is activated but it won't let any memory to be allocated, forcing VCPI apps to allocate via XMS.
* planar VGA modes work, at least for now, via coalesced IO, a KVM feature which puts mmio/port writes in a buffer which can later be processed. This is much faster than a VMEXIT for every i/o or write but quite a bit slower than instremu (as they are still trapped inside the VM). Using instremu would imply teaching it about paging.
* works with FDPP now
* doom/duke3d/jazz work and can be played, but sound is choppy.

stsp commented:
Interesting.
So I suppose, at least from a configuration
POV, only $_ems prevents this from merging
in? It already works with default $_dpmi
setting and with fdpp, so that sounds good.
Will it not require lowering $_ems in the future?

stsp commented:
> planar VGA modes work, at least for now, via coalesced IO, a KVM feature which puts mmio/port writes
> in a buffer which can later be processed.

This is cool but IIRC doom (and most
other planar-mode games) read from
mmio just as frequent as they write.
So batching doesn't actually work, or
does it?

bartoldeman committed:
VCPI: only works with full KVM

Otherwise report "unsupported" for int67/ah=de. This should fix CI.

bartoldeman commented:
I'll double check but IIRC writes are much more frequent than reads to planar VGA memory.

I'm working on eliminating the EMS requirement, by implementing interfaces int67 0xde03/de04/de05 properly (they can just get some extmem/XMS, no need to alias map, it's just to get the accounting right, since programs expect to be able to allocate pages via those interfaces if EMS int67 ah=0x42 reports >0 memory).

JEMM/EMM386 have a NOEMS switch which is emulated with $_ems = (4), since EMS pages are 16k this gets rounded to 0 but better not to need that hack with the above.

stsp commented:
> I'll double check but IIRC writes are much more frequent than reads to planar VGA memory.

I was thinking about batching on top
of instremu. In theory, if you can create
the large enough write batches, then you
can submit them to the different thread.
But my conclusion was that you need
really large batches for this to make any
effect, and frequent reading was ruining
that idea.
But maybe your findings would be
different. Overall, finding a way to speed
up planar modes, even mildly, would be
very cool. Currently, I am afraid, we have
the slowest planar mode emulation among
any competitors, including dosemu1. :(

bartoldeman committed:
VCPI: implement memory allocation calls.

Applies the page allocator to the physical memory area starting at 16M up to
16M + $_ems. To do: deal with DPMI overlaps if $_ems > 16M.
This removes all special requirements on dosemu.conf parameters for using
VCPI.

bartoldeman commented:
QEMU with KVM uses coalesced MMIO/PIO for planar VGA from what I could see, though it's likely faster in TCG mode for that. Will still look at that later.

I've implemented the memory allocation functions now; they operate on the memory area at 16M, so there is no overlap with XMS and EXTMEM. I should still add a configuration parameter which forcibly disables VCPI, and deal with overlaps for VCPI/DPMI memory, and add checks ensuring VCPI/DPMI are never running at the same time. But it's getting much closer now.

stsp commented:
Sound cool, thanks!
Can you check that the Origin games
are running? This is the primary (perhaps
the only) motivator for having vcpi.

stsp commented:
> QEMU with KVM uses coalesced MMIO/PIO for planar VGA from what I could see, though it's likely
> faster in TCG mode for that.

Possibly, but here planar modes are
even more slow with simx86... So with
planar modes we suck on all fronts,
including jit.

bartoldeman commented:
I did a little experiment with
```
diff --git a/src/base/dev/vga/vgaemu.c b/src/base/dev/vga/vgaemu.c
index 952086685..c63d95edb 100644
--- a/src/base/dev/vga/vgaemu.c
+++ b/src/base/dev/vga/vgaemu.c
@@ -786,8 +786,11 @@ int vga_access(dosaddr_t r, dosaddr_t w)
 
 unsigned char vga_read(dosaddr_t addr)
 {
+  static unsigned int reads = 0;
   if (!vga.inst_emu || !vga_read_access(addr))
     return READ_BYTE(addr);
+  reads++;
+  if ((reads % 10000) == 0) printf("reads=%d\n", reads);
   return Logical_VGA_read(addr - vga.mem.bank_base);
 }
 
@@ -827,6 +830,7 @@ void vga_mark_dirty(dosaddr_t vga_addr, int len)
 
 void vga_write(dosaddr_t addr, unsigned char val)
 {
+  static unsigned int writes = 0;
   if (!vga.inst_emu || !vga_bank_access(addr)) {
     addr = vga_get_mem_base_offset(addr);
     if (addr != (dosaddr_t)-1) {
@@ -835,6 +839,8 @@ void vga_write(dosaddr_t addr, unsigned char val)
     }
     return;
   }
+  writes++;
+  if ((writes % 100000) == 0) printf("writes=%d\n", writes);
   Logical_VGA_write(addr - vga.mem.bank_base, val);
 }
 
```
and DOOM, reads were only at startup, going to 140,000, writes kept increasing after, a minute or so in at 17,100,000.
DOOM is a bit of an odd case also, since it is playable without instremu, just then some of the detailed animations are missing, basically regular 320x200 mode int13h but with mode-X special effects.

Jazz Jackrabbit is more balanced, indeed at writes=7,400,000 reads=3,980,000, so only a ratio 2:1.

Good old bgidemo, my original instremu testcase from almost 25 years ago is at
writes=4,700,000,  reads=1,620,000.

For the JIT I can see some ways to speed it up, basically it goes via cpatch, so no faults, but the cpatch code itself doesn't use `unprotected_page_cache` so it still has too much overhead.

Replacing instremu by JIT is one of those very long time ideas that never came to be so far.

stsp commented:
Even speeding that on jit would
be great. Eventually we need to
be able to quickly switch between
jit and kvm to have a good performance.
This is how we currently temporary
switch to instremu after a fault in
a hope there would be more faults.
But with jit this can be much more
controllable, as there you can only
compile the code that has vga
accesses, and quickly switch to
KVM when no more compiled code
found.
There can even be a mixed mode,
where you run kvm if no compiled
code is found, and jit compiles it in
a separate thread in between. :)

stsp commented:
I think to get the best performance, we
first need to unbind the interp from execution
flow, so that it only produces code nodes
in a thread. This doesn't even look very
difficult to do.
The next step can be to embed kvm as
an accellerator there, such as when you
have code tree miss, you run kvm while
interp keeps jitting.
Do you think such architecture will be
good performance-wise?
Those planar mode problems will simply
not exist there.

bartoldeman commented:
I had a quick look with perf and there was one piece of low-hanging fruit with the simulator (#1912).
JIT and KVM did not show anything obvious. I did run with speed (hogthreshold) 0. I'll keep looking.

bartoldeman committed:
VCPI: add config.

$_vcpi = (0) by default, any other value for now means "on" but in future
some value > $_ems can denote higher amounts of VCPI memory than EMS memory.

bartoldeman committed:
VCPI: add support for getting and setting debug registers.

bartoldeman committed:
VCPI: alias map high VCPI pages for EMS

Some VCPI applications rely on page table mappings between the EMS page
frame and high memory, so we must set them up and release them on demand
as alias mappings.

bartoldeman commented:
After doing high EMS mappings on demand, Crusader Demo works now, though only in FreeDOS; with FDPP there are some XMS issues, so still figuring that out.
![Screenshot_20230221_080905](https://user-images.githubusercontent.com/10634630/220355890-38db3a6c-7e1d-4d03-8bac-0f0dc7930a77.png)

stsp commented:
Can there be this problem:
https://github.com/dosemu2/dosemu2/issues/875#issuecomment-1424295164
fdpp moves things to UMB/HMA
very aggressively, so try `$_dos_up = (0)`.

bartoldeman commented:
It's an issue with the internal XMS driver. For some reason (may just be something odd in my config), `device=himemx.exe` in `config.sys` gets ignored with FDPP. If I comment it out explicitly with FreeDOS, I get the same error (`ERROR: error allocating 7273472 bytes for xms`).

stsp commented:
You need to create userhook.sys
instead of config.sys.

bartoldeman commented:
it looks like `userhook.sys` comes too late, but adding
`device=himemx.exe`
just before `device=dosemu\umb.sys` in `fdppconf.sys`, in combination with `$_dos_up=(0)` makes crusader work, thanks!

stsp commented:
Thanks for a note.
I moved globhook.sys earlier, so
it is now possible to load himem.sys
from there instead of userhook.sys.

What is the reason of needing an
external himem.sys though? Or is
that still a mystery?

stsp commented:
Btw, there is also a `$_dos_up=(2)` mode
where fdpp moves itself to HMA instead of
UMB. And that one is incompatible with
external himem as it would trash HMA on
load and trash fdpp that is already there.
But you may still find it interesting to take
a look at how much memory can be freed
by the very aggressive HMA usage. :)
FreeDOS doesn't do remotely as good.

bartoldeman commented:
It's still a mystery, I was thinking of the 16MB barrier, since, if $_ext_mem + $_xms > 15MB I get this error message:
`ERROR: error allocating 1018880 bytes for xms` which comes from `map_EMB` because `pgaalloc` fails.
(this is for $_xms=$ext_mem=8192) but if I then decrease $_xms to 6072 I get this:
```
ERROR: MCB corruption, good:%P bad:%P

ERROR: fdpp: abort at hmamgr.cc:331
```

bartoldeman commented:
FreeDOS crashes with internal XMS only if xms+extmem > 15MB, otherwise it works (no MCB corruption issue). So it looks like there are two issues here.

stsp commented:
Sounds like I need to reproduce the
fdpp problem (its not like you can't
debug fdpp with gdb, but its not a quick
start), so any short instructions to
recreate your setup?

bartoldeman commented:
oh I was too impatient and aborted. So the map_EMB error is just noisy and it works with FreeDOS for the default $_xms too.

That leaves the MCB error for FDPP, which happens independent of $_xms settings.

so basically I downloaded cru-dem.zip from https://www.dosgamesarchive.com/file/crusader-no-remorse/cru-dem

then you can start crusader.exe from there but you need to do something like
lredir -f -C d: linux\fs/path/to/cd.rom
to mount the cd.rom directory in the unpacked zip as drive d:.

stsp commented:
But this needs entire patch-set and
disabling dpmi, or will it corrupt anyway?

bartoldeman commented:
yes you do need to check out this branch (or apply 
https://patch-diff.githubusercontent.com/raw/dosemu2/dosemu2/pull/1880.diff
if that's quicker) No need to disable dpmi, but you do need to set
$_vcpi = (auto).

stsp commented:
Thanks, reproduced.
Even w/o CD dances.

> if that's quicker)

I have this in my .git/config:
`fetch = +refs/pull/*/head:refs/remotes/origin/pr/*`
which automatically fetches the PRs.
So that's quick enough. :)

stsp commented:
With this patch:
```
--- a/src/base/emu-i386/kvm.c
+++ b/src/base/emu-i386/kvm.c
@@ -1050,7 +1050,9 @@ static unsigned int kvm_run(void)
   }
 
   while (!exit_reason) {
+warn("enter %c\n", lowmem_base[(0xffff<<4)+0x10]);
     int ret = ioctl(vcpufd, KVM_RUN, NULL);
+warn("exit %c\n", lowmem_base[(0xffff<<4)+0x10]);
     int errn = errno;
 
     /* KVM should only exit for four reasons:
```

I eventually see in the log:
```
Warning: enter M
Warning: exit .
```

So the game corrupts HMA and freedos
is any "better" only because it doesn't have
the proper HMA manager.

stsp commented:
CS is 0x2a8 on enter and mcbs are:
```
ADDR(LOW) PARAS  OWNER
0263:0000 0x0013 [CRUSADER - Environment]
0277:0000 0x0770 [CRUSADER]
09e8:0000 0x0100 [9ced - Data]
```
So its a crusader's CS.

bartoldeman commented:
ok, this could be an A20-line issue then. I think perhaps some place needs to force a20 on to make this work.

bartoldeman commented:
The dosbox instructions have some loadfix stuff in them, so that may be it!

stsp commented:
OK I see the bug actually... or
a part of it. Give me a bit of time.

stsp commented:
Should now be fixed.
The problem is that our debugger
queried the HMA start, and fdpp is
too permissive and just enables HMA
if it wasn't enabled already.
Its undesirable when the program
wants to use HMA on its own as
crusader does.
So I added an extra check to the debugger.
Could also restrict fdpp, but oh well.

Note that with `DOS=HIGH` the problem
will re-appear. But who cares if its not
a default. :)

bartoldeman commented:
Yes it works now, thanks, just with noisy error messages when internal XMS is used.
Also the DOSEMU2 window title bar now changed CRUSADER to some strange characters, but will try to narrow down when that started. It's all consistent between FDPP and fdkernel though now!

bartoldeman committed:
VCPI: pagetable entries for EMS pages need 4 entries!

bartoldeman committed:
KVM: VCPI: replace outdated DIRTY LOG enabler by MMIO enabler.

When entering VCPI, we need to convert PROT_NONE maps into MMIO maps,
dirty logging is always ok.
